### ISSUE TITLE
[feature] Learning about failures between channels

### ISSUE THREAD

**Issue by ziggie1984** [rkfg/regolancer#19]
When using regolancer, we fetch all possible routes a specific channel combination has before trying all routes regarding this channel-combination. Problem with this approach is that our system does not learn about the channel failure which would give us more information and discard maybe routes which have the same channels which failed before. Hence making regolancer more efficient.
Do you know whether SendPaymentV2 would solve this kind of problems because we let lnd do the job, gaining a better overview of the channel graph and we would just give lnd the outgoing channel and the destination ? Otherwise we could build up those information locally, what do you think?

**Comment by rkfg** [comment]
Hmm? Not sure what you mean, if a there's no routes between a pair of channels (source-target) this pair moved to the failure cache and this channel pair isn't considered again until this pair times out (currently the timeout is 5 minutes). lnd currently always returns just one route so we shouldn't waste any resources on this. And if the route fails this data is stored in its Mission Control so it shouldn't propose the same route again.

**Comment by ziggie1984** [comment]
Hmm got this from a noderunner, and that kind of irritated me, but I checked, really lnd does only return 1 route for every query. So might be that lnd fails to gain failure information bc. normally it should not use the second one again ?

<img width="1364" alt="Screenshot 2022-10-28 at 20 52 56" src="https://user-images.githubusercontent.com/90319354/198711360-7dd75a31-6a42-4941-affe-c25cd2d6e61f.png">

**Comment by rkfg** [comment]
Could be an lnd bug, yep. What's it version? You can query the mission control info with `lncli querymc`, see what data it has for that channel. I checked my recent rebalance and also easily found such an issue though it didn't happen twice in a row, there were other attempts between them:
```
2022/10/28 21:00:11 Attempt #1899, amount: 100000 (max fee: 2 sat | 26 ppm )                                                                                                          
Total fee: 1 sat | 15 ppm                                                                                                                                                             
819839850139418625        [lnmarkets.com|269ch|27.17,966,169sat|03271338633d2d37b285dae4df40b413d8c6c791fbee7797bc5dc70812196d7d5c]
783895715608788992 1100   [Bitcoin Martian|36ch|1.81,399,998sat|03f20b295a728be64537bd7df1bf9092ed4bbf09b2c10f68c482d6fcb4f2232b28]
835770674335907841 100    [deezy.io ⚡✨|903ch|113.96,452,286sat|024bfaf0cabe7f874fd33ebf7c6f4e5385971fc504ef3f492432e9e3ec77e1b5cf]
833462799225061376 0      [Izanagi🚹|24ch|1.80,000,000sat|039b18a4c32a63d7d390c4fe61b26598fc9aad2ce19303ca5fec01f10687d154e8]
833770662516293632 400    [rkfg|64ch|2.34,955,783sat|021c3ec6432d2b9b5abcb01dd64b3a8f2afe3ba7d8c021f63ffd0a994cd3bc9b88]

2022/10/28 21:00:12 TEMPORARY_CHANNEL_FAILURE lnmarkets.com ⇒ Bitcoin Martian                                                                                                         

...

2022/10/28 21:01:50 Attempt #1914, amount: 100000 (max fee: 3 sat | 36 ppm )                                                                                                          
Total fee: 1 sat | 11 ppm                                                                                                                                                             
819839850139418625        [lnmarkets.com|269ch|27.17,966,169sat|03271338633d2d37b285dae4df40b413d8c6c791fbee7797bc5dc70812196d7d5c]
783895715608788992 1100   [Bitcoin Martian|36ch|1.81,399,998sat|03f20b295a728be64537bd7df1bf9092ed4bbf09b2c10f68c482d6fcb4f2232b28]
835770674335907841 100    [deezy.io ⚡✨|903ch|113.96,452,286sat|024bfaf0cabe7f874fd33ebf7c6f4e5385971fc504ef3f492432e9e3ec77e1b5cf]
811220778577297409 0      [rkfg|64ch|2.34,955,783sat|021c3ec6432d2b9b5abcb01dd64b3a8f2afe3ba7d8c021f63ffd0a994cd3bc9b88]

2022/10/28 21:01:53 TEMPORARY_CHANNEL_FAILURE lnmarkets.com ⇒ Bitcoin Martian                                                                                                         
```
Interesting indeed. One way to fix it without going deep into lnd's internals is to add failed channels to the excluded ones, it would inflate RPC requests fairly quick I think. To debug we might add some code to query the MC. Unfortunately, there's no way to filter the result without receiving it first so it would be slow. Then we can check if MC has indeed recorded the failure and also store the chan id that failed (in a simple map with uint64 keys). If we encounter it again we'd need to compare the MC result that's related to this channel with whatever there was the last time. If they're the same then the channel was selected without MC consideration and it might be a bug, a misconfiguration or a misunderstanding.

**Comment by rkfg** [comment]
Couldn't find any issues about it in the lnd repo. Need to do our own experiments, then open an issue if it's easy to reproduce. Maybe build a simple test case in Polar (not using regolancer but a simple test program that connects to lnd, queries a route and tries a payment.

**Comment by ziggie1984** [comment]
Ok thanks for your inputs, will write some tests tomorrow with polar

**Comment by rkfg** [comment]
After some experiments it turned out this behavior of mission control is correct. The amount on every hop includes fees for the next hops so in the first case (in both examples above) the total amount including fees was higher than in the second case. MC doesn't have any tolerance settings and it allows a channel to be chosen if the amount is even 1 msat less than the known failed amount. In majority of cases it doesn't make sense to try, say, 100 052 sat payment if 100 053 is known to fail. We're trying to make such a heuristic optimization in #23.

**Comment by rkfg** [comment]
Fixed by #23
